src/odemis/acq/align/z_localization.py

```diff
@@ -0,0 +1,193 @@
+    a = calibration_data["a"]
+    b = calibration_data["b"]
+    c = calibration_data["c"]
+    d = calibration_data["d"]
+    return a * (z - c) ** 2 + d * (z - c) ** 3 + b
+
+
+def solve_psf(z, obs_x, obs_y, calibration_data, model_function=huang):
+    """
+    Function used with fmin_cg(scipy) to get the least squares error for the z position functions (huang/thunderstorm)
+
+    :param z (float): z position of the feature
+    :param obs_x (float): Observed sigma_x, size of the feature in x direction.
+    :param obs_y (float): Observed sigma_y, size of the feature in y direction.
+    :param calibration_data (dict): contains the constants in both x and y direction for equation with the required keys
+    :param model_function (func): A function that describes the sigma_x/sigma_y as function of z. (huang/thunderstorm)
+    :return (float): least squares error for the model_function in both x and y direction
+    """
+    cal_x = model_function(z, calibration_data["x"])
+    cal_y = model_function(z, calibration_data["y"])
+    return (obs_x ** 0.5 - cal_x ** 0.5) ** 2 + (obs_y ** 0.5 - cal_y ** 0.5) ** 2
+
+# TODO K.K. reconsider the z_stack_step_size as argument since it is only used for the initial guess of the fit.
+def determine_z_position(image, calibration_data, fit_tol=0.1, z_stack_step_size=50e-9):
+    """
+    Function to determine the z position of feature in an image that was taken with a lens with astigmatism and
+    corresponding calibration data. Via a Gaussian fit the width and height of the feature are determined. Using a
+    fit on the equation of Huang the z position is then approximated. The function includes various warning flags
+    that may be raised when the calculation seems to give inaccurate results.
+
+    :param image (numpy.array): 2d array containing only the feature to be analyzed.
+    :param calibration_data (dict): contains the data from the calibration performed using the jupyter notebook with the
+                                    following keys:
+                                    x (dict) --> a, b ,c, d, w0 (floats), fit on the equation of Huang in x direction
+                                    y (dict) --> a, b ,c, d, w0 (floats), fit on the equation of Huang in y direction
+                                    feature_angle (float) angle of the ellipsoidal shapes w.r.t. the positive Y axis (anti-clockwise is positive)
+                                    upsample_factor (int) number with which the data is up sampled during the calibration
+                                    z_least_confusion (float) location in z in meters where the least confusion is present in the image (in focus)
+                                    z_calibration_range (tuple) --> (min, max) floats of the min and max z value in meters w.r.t. the z_least_confusion
+
+    :param z_stack_step_size (float): the step size in meters between two images in the z stack
```

Comment:
Link the answer in this conversation:
https://github.com/delmic/odemis/pull/2076#discussion_r836118719

> I would say looking at the step size is better. Otherwise the precision of your initial guess is dependent on your max_range. That feels a bit odd.
Though the 'random value" I like. Maybe I was already thinking about removing the z_stack_step_size keyword argument. Maybe I should just hardcode it and comment on it?